core/jobs/batch_jobs/story_node_jobs.py

```diff
@@ -0,0 +1,183 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jobs used for populating story node."""
+
+from __future__ import annotations
+
+import logging
+
+from core import utils
+from core.jobs import base_jobs
+from core.jobs.io import ndb_io
+from core.jobs.transforms import job_result_transforms
+from core.jobs.types import job_run_result
+from core.platform import models
+
+import apache_beam as beam
+import result
+from typing import Tuple
+
+MYPY = False
+if MYPY: # pragma: no cover
+    from mypy_imports import datastore_services
+    from mypy_imports import story_models
+    from mypy_imports import topic_models
+
+(story_models, topic_models) = models.Registry.import_models([
+    models.Names.STORY, models.Names.TOPIC])
+datastore_services = models.Registry.import_datastore_services()
+
+
+class PopulateStoryNodeJob(base_jobs.JobBase):
+    """Job that migrates story nodes, for serial chapter launch feature"""
+
+    DATASTORE_UPDATES_ALLOWED = True
+
+    def _update_story_node(
+            self, story_model: story_models.StoryModel
+    ) -> result.Result[
+        story_models.StoryModel,
+        Tuple[str, Exception]
+    ]:
+        """Populate the 5 new fields in each story node of the StoryModel
+        instance, namely: status, planned_publication_date_msecs,
+        last_modified_msecs, first_publication_date_msecs, unpublishing_reason.
+
+        Args:
+            story_model: StoryModel. The story whose nodes have to be
+                populated.
+
+        Returns:
+            Result(StoryModel, (str, Exception)). Result containing the
+            updated StoryModel instance to be put.
+        """
+
+        nodes = story_model.story_contents['nodes']
+
+        try:
+            with datastore_services.get_ndb_context():
+                topic_model = topic_models.TopicModel.get(
+                    story_model.corresponding_topic_id)
+                story_reference = next(
+                    story_ref for story_ref in (
+                        topic_model.canonical_story_references)
+                        if story_ref['story_id'] == story_model.id)
+                for node in nodes:
+                    node['unpublishing_reason'] = None
+                    node['status'] = 'Draft'
+                    if story_reference['story_is_published']:
+                        node['status'] = 'Published'
+
+                    current_topic_version = topic_model.version
+                    story_published_on = None
+                    for version in range(current_topic_version, 0, -1):
+                        snapshot_id = topic_model.get_snapshot_id(
+                            topic_model.id, version)
+                        topic_metadata = (
+                            topic_models.TopicSnapshotMetadataModel.get(
```

Comment:
Is it possible to avoid making so many get requests in the datastore? (I'm assuming we are making ~n^2 requests here)

Reply:
Yes I understand. I actually had given much thought to optimize this and also discussed this with Hitesh, but this was the best we could decide. Because for each node of the story, we need to traverse back the topic version history as well as story version history to get the relevant information for the node.
I thought that we had 2 options: 
1. Either to do like this and continue fetching the previous version models till the requirement is met and break the loop
2. Or to fetch all the versions of the topic and that story at the start, and then loop till we get the relevant data and break the loop. However doing this way fetches even those earlier versions which might never be needed at all for any of the nodes, and also maintaining the PCollection array of topic version models and story version models for each topic and story also appears to be a complex data structure, which looks difficult to work with.

So for simplicity and clear understandablity of the code, I have written like this.

If you have any alternative ideas to implement this, please share a description so that we can work on it.